// remult/esm/server/remult-admin.js
import { getRelationFieldInfo, } from '../src/remult3/relationInfoMember.js';
import { Filter } from '../src/filter/filter-interfaces.js';
import { getHtml } from './get-remult-admin-html.js';
import { getValueList } from '../src/remult3/RepositoryImplementation.js';
import { InputTypes } from '../inputTypes.js';
export default function remultAdminHtml(options) {
    const { head, rootPath, requireAuthToken, disableLiveQuery } = options;
    return getHtml()
        .replace('<!--PLACE_HERE_HEAD-->', head)
        .replace('<!--PLACE_HERE_BODY-->', `<script>
  window.optionsFromServer = ${JSON.stringify({
        rootPath,
        requireAuthToken,
        disableLiveQuery,
    })}
</script>`);
}
export function buildEntityInfo(options) {
    const entities = [];
    for (const metadata of options.entities.map((e) => options.remult.repo(e).metadata)) {
        let fields = [];
        let relations = [];
        let ids = {};
        for (const f of metadata.idMetadata.fields) {
            ids[f.key] = true;
        }
        for (const x of metadata.fields.toArray()) {
            if (!x.includedInApi(undefined))
                continue;
            try {
                let relation;
                let valFieldKey = x.key;
                const info = getRelationFieldInfo(x);
                if (info) {
                    const relInfo = info.getFields();
                    const relRepo = options.remult.repo(info.toEntity);
                    const where = typeof info.options.findOptions === 'object' &&
                        info.options.findOptions.where
                        ? Filter.entityFilterToJson(relRepo.metadata, info.options.findOptions.where)
                        : undefined;
                    const idField = relRepo.metadata.idMetadata.field.key;
                    if (info.type === 'reference' || info.type === 'toOne') {
                        if (info.type == 'toOne') {
                            for (const key in relInfo.fields) {
                                if (Object.prototype.hasOwnProperty.call(relInfo.fields, key)) {
                                    const element = relInfo.fields[key];
                                    valFieldKey = element;
                                }
                            }
                        }
                        if (relRepo.metadata.apiReadAllowed) {
                            relation = {
                                ...relInfo,
                                where,
                                entityKey: relRepo.metadata.key,
                                idField,
                                captionField: relRepo.metadata.fields
                                    .toArray()
                                    .find((x) => x.key != idField && x.valueType == String)?.key,
                            };
                        }
                    }
                    else if (info.type === 'toMany') {
                        if (relRepo.metadata.apiReadAllowed) {
                            relations.push({
                                ...relInfo,
                                where,
                                entityKey: relRepo.metadata.key,
                                key: x.key,
                                caption: x.label,
                            });
                        }
                        continue;
                    }
                }
                fields.push({
                    key: x.key,
                    readOnly: !x.apiUpdateAllowed(),
                    values: getValueList(x),
                    valFieldKey,
                    caption: x.label,
                    relationToOne: relation,
                    inputType: x.inputType,
                    type: x.inputType === InputTypes.json
                        ? 'json'
                        : x.valueType === Number
                            ? 'number'
                            : x.valueType === Boolean
                                ? 'boolean'
                                : x.valueType === Date
                                    ? 'date'
                                    : 'string',
                });
            }
            catch (error) {
                console.error(`[remult-admin] Error with ${metadata.key}.${x.key} field.`);
                console.error(`[remult-admin]`, error);
            }
        }
        if (metadata.apiReadAllowed) {
            let superKey = metadata.key;
            let caption = metadata.label;
            const nbOfEntities = entities.filter((e) => e.key === metadata.key).length;
            if (nbOfEntities > 0) {
                superKey = metadata.key + '_ext_' + nbOfEntities;
                caption = metadata.label + '*'.repeat(nbOfEntities);
            }
            entities.push({
                superKey,
                key: metadata.key,
                caption,
                ids,
                fields,
                relations,
                defaultOrderBy: metadata.options.defaultOrderBy ?? { id: 'asc' },
            });
        }
    }
    return entities;
}
