// remult/server/remult-admin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = remultAdminHtml;
exports.buildEntityInfo = buildEntityInfo;
var tslib_1 = require("tslib");
var relationInfoMember_js_1 = require("../src/remult3/relationInfoMember.js");
var filter_interfaces_js_1 = require("../src/filter/filter-interfaces.js");
var get_remult_admin_html_js_1 = require("./get-remult-admin-html.js");
var RepositoryImplementation_js_1 = require("../src/remult3/RepositoryImplementation.js");
var inputTypes_js_1 = require("../inputTypes.js");
function remultAdminHtml(options) {
    var head = options.head, rootPath = options.rootPath, requireAuthToken = options.requireAuthToken, disableLiveQuery = options.disableLiveQuery;
    return (0, get_remult_admin_html_js_1.getHtml)()
        .replace('<!--PLACE_HERE_HEAD-->', head)
        .replace('<!--PLACE_HERE_BODY-->', "<script>\n  window.optionsFromServer = ".concat(JSON.stringify({
        rootPath: rootPath,
        requireAuthToken: requireAuthToken,
        disableLiveQuery: disableLiveQuery,
    }), "\n</script>"));
}
function buildEntityInfo(options) {
    var e_1, _a;
    var _b, _c;
    var entities = [];
    var _loop_1 = function (metadata) {
        var e_2, _f, e_3, _g;
        var fields = [];
        var relations = [];
        var ids = {};
        try {
            for (var _h = (e_2 = void 0, tslib_1.__values(metadata.idMetadata.fields)), _j = _h.next(); !_j.done; _j = _h.next()) {
                var f = _j.value;
                ids[f.key] = true;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_j && !_j.done && (_f = _h.return)) _f.call(_h);
            }
            finally { if (e_2) throw e_2.error; }
        }
        var _loop_2 = function (x) {
            if (!x.includedInApi(undefined))
                return "continue";
            try {
                var relation = void 0;
                var valFieldKey = x.key;
                var info = (0, relationInfoMember_js_1.getRelationFieldInfo)(x);
                if (info) {
                    var relInfo = info.getFields();
                    var relRepo = options.remult.repo(info.toEntity);
                    var where = typeof info.options.findOptions === 'object' &&
                        info.options.findOptions.where
                        ? filter_interfaces_js_1.Filter.entityFilterToJson(relRepo.metadata, info.options.findOptions.where)
                        : undefined;
                    var idField_1 = relRepo.metadata.idMetadata.field.key;
                    if (info.type === 'reference' || info.type === 'toOne') {
                        if (info.type == 'toOne') {
                            for (var key in relInfo.fields) {
                                if (Object.prototype.hasOwnProperty.call(relInfo.fields, key)) {
                                    var element = relInfo.fields[key];
                                    valFieldKey = element;
                                }
                            }
                        }
                        if (relRepo.metadata.apiReadAllowed) {
                            relation = tslib_1.__assign(tslib_1.__assign({}, relInfo), { where: where, entityKey: relRepo.metadata.key, idField: idField_1, captionField: (_b = relRepo.metadata.fields
                                    .toArray()
                                    .find(function (x) { return x.key != idField_1 && x.valueType == String; })) === null || _b === void 0 ? void 0 : _b.key });
                        }
                    }
                    else if (info.type === 'toMany') {
                        if (relRepo.metadata.apiReadAllowed) {
                            relations.push(tslib_1.__assign(tslib_1.__assign({}, relInfo), { where: where, entityKey: relRepo.metadata.key, key: x.key, caption: x.label }));
                        }
                        return "continue";
                    }
                }
                fields.push({
                    key: x.key,
                    readOnly: !x.apiUpdateAllowed(),
                    values: (0, RepositoryImplementation_js_1.getValueList)(x),
                    valFieldKey: valFieldKey,
                    caption: x.label,
                    relationToOne: relation,
                    inputType: x.inputType,
                    type: x.inputType === inputTypes_js_1.InputTypes.json
                        ? 'json'
                        : x.valueType === Number
                            ? 'number'
                            : x.valueType === Boolean
                                ? 'boolean'
                                : x.valueType === Date
                                    ? 'date'
                                    : 'string',
                });
            }
            catch (error) {
                console.error("[remult-admin] Error with ".concat(metadata.key, ".").concat(x.key, " field."));
                console.error("[remult-admin]", error);
            }
        };
        try {
            for (var _k = (e_3 = void 0, tslib_1.__values(metadata.fields.toArray())), _l = _k.next(); !_l.done; _l = _k.next()) {
                var x = _l.value;
                _loop_2(x);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_l && !_l.done && (_g = _k.return)) _g.call(_k);
            }
            finally { if (e_3) throw e_3.error; }
        }
        if (metadata.apiReadAllowed) {
            var superKey = metadata.key;
            var caption = metadata.label;
            var nbOfEntities = entities.filter(function (e) { return e.key === metadata.key; }).length;
            if (nbOfEntities > 0) {
                superKey = metadata.key + '_ext_' + nbOfEntities;
                caption = metadata.label + '*'.repeat(nbOfEntities);
            }
            entities.push({
                superKey: superKey,
                key: metadata.key,
                caption: caption,
                ids: ids,
                fields: fields,
                relations: relations,
                defaultOrderBy: (_c = metadata.options.defaultOrderBy) !== null && _c !== void 0 ? _c : { id: 'asc' },
            });
        }
    };
    try {
        for (var _d = tslib_1.__values(options.entities.map(function (e) { return options.remult.repo(e).metadata; })), _e = _d.next(); !_e.done; _e = _d.next()) {
            var metadata = _e.value;
            _loop_1(metadata);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return entities;
}
